// remote-redux-devtools-on-debugger/lib/main.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _injectDebugger = require('./injectDebugger');

var injectDebugger = _interopRequireWildcard(_injectDebugger);

var _injectServer = require('./injectServer');

var injectServer = _interopRequireWildcard(_injectServer);

var _runServer = require('./runServer');

var _runServer2 = _interopRequireDefault(_runServer);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var name = 'react-native';
var bundleCode = _fs2.default.readFileSync(_path2.default.join(__dirname, '../bundle.js'), 'utf-8');

var getModulePath = function getModulePath(moduleName) {
  return _path2.default.join(process.cwd(), 'node_modules', moduleName);
};

var log = function log(pass, msg) {
  var prefix = pass ? _chalk2.default.green.bgBlack('PASS') : _chalk2.default.red.bgBlack('FAIL');
  var color = pass ? _chalk2.default.blue : _chalk2.default.red;
  console.log(prefix, color(msg));
};

var getFullPath = function getFullPath(filePath) {
  return _path2.default.resolve(process.cwd(), filePath);
};

var assignSecureOptions = function assignSecureOptions(options, _ref) {
  var secure = _ref.secure;
  var key = _ref.key;
  var cert = _ref.cert;
  var passphrase = _ref.passphrase;
  return _extends({}, options, secure ? {
    protocol: 'https',
    key: key ? getFullPath(key) : null,
    cert: cert ? getFullPath(cert) : null,
    passphrase: passphrase || null
  } : null);
};

module.exports = function (argv) {
  var modulePath = getModulePath(argv.desktop ? 'react-native-desktop' : name);

  // Revert all injection
  if (argv.revert) {
    var passServ = injectServer.revert(modulePath);
    var msg = 'Revert injection of RemoteDev server from React Native local server';
    log(passServ, msg + (!passServ ? ', the file \'' + injectServer.fullPath + '\' not found.' : '.'));

    var passDbg = injectDebugger.revert(modulePath);
    msg = 'Revert injection of RemoteDev monitor from React Native debugger';
    log(passDbg, msg + (!passDbg ? ', the file \'' + injectDebugger.fullPath + '\' not found.' : '.'));

    return passServ && passDbg;
  }

  var options = argv.hostname || argv.port ? {
    secure: argv.secure,
    hostname: argv.hostname || 'localhost',
    port: argv.port || 8000
  } : null;

  // Inject server
  if (argv.injectserver) {
    var pass = injectServer.inject(modulePath, assignSecureOptions(options, argv));
    var _msg = 'Inject RemoteDev server into React Native local server';
    log(pass, _msg + (pass ? '.' : ', the file \'' + injectServer.fullPath + '\' not found.'));
    if (!pass) return false;
  }

  // Inject debugger
  if (argv.injectdebugger) {
    var _pass = injectDebugger.inject(modulePath, bundleCode, options);
    var _msg2 = 'Inject RemoteDev monitor into React Native debugger';
    log(_pass, _msg2 + (_pass ? '.' : ', the file \'' + injectDebugger.fullPath + '\' not found.'));
    if (!_pass) return false;
  }

  // Run RemoteDev server
  if (argv.runserver) {
    return (0, _runServer2.default)(assignSecureOptions(options || { secure: argv.secure, port: 8000 }, argv));
  }
  return true;
};
